components/mediation/data-publishers/org.wso2.micro.integrator.observability/src/main/java/org/wso2/micro/integrator/observability/handler/metricHandler/MetricHandler.java

```diff
@@ -0,0 +1,443 @@
+    }
+
+    /**
+     * Increment the error request count received by an api.
+     *
+     * @param name             The metric name
+     * @param apiInvocationUrl api Invocation URL
+     */
+    private void incrementApiErrorCount(String name, String apiInvocationUrl) {
+        Map<String, String[]> metricCountMap = new HashMap();
+        metricCountMap.put(MetricConstants.API_REQUEST_COUNT_ERROR_TOTAL, new String[]{name,
+                SynapseConstants.FAIL_SAFE_MODE_API, apiInvocationUrl});
+        metricReporterInstance.incrementCount(MetricConstants.API_REQUEST_COUNT_ERROR_TOTAL, metricCountMap);
+    }
+
+    /**
+     * Increment the error request count received by an inbound endpoint.
+     *
+     * @param name The metric name
+     */
+    private void incrementInboundEndpointErrorCount(String name) {
+        Map<String, String[]> metricCountMap = new HashMap();
+        metricCountMap.put(MetricConstants.INBOUND_ENDPOINT_REQUEST_COUNT_ERROR_TOTAL, new String[]{name,
+                MetricConstants.INBOUND_ENDPOINT});
+        metricReporterInstance.incrementCount(MetricConstants.INBOUND_ENDPOINT_REQUEST_COUNT_ERROR_TOTAL,
+                metricCountMap);
+    }
+
+    /**
+     * Get the api name.
+     *
+     * @param contextPath The api context path
+     * @param synCtx The Synapse Message Context
+     * @return String The api name
+     */
+    private String getApiName(String contextPath, MessageContext synCtx) {
+        String apiName = null;
+        for (API api : synCtx.getEnvironment().getSynapseConfiguration().getAPIs()) {
+            if (RESTUtils.isRESTApiPath(contextPath, api.getContext())) {
+                apiName = api.getName();
+                if (api.getVersionStrategy().getVersion() != null && !"".equals(api.getVersionStrategy().
+                        getVersion())) {
+                    apiName = apiName + ":v" + api.getVersionStrategy().getVersion();
+                }
+                synCtx.setProperty(RESTConstants.IS_PROMETHEUS_ENGAGED, true);
+                synCtx.setProperty(RESTConstants.PROCESSED_API, api);
+            }
+        }
+        return apiName;
+    }
+
+    /**
+     * No Synapse Exception is thrown because the mediation flow should not be affected by
+     * an exception thrown in the Observability Handler.
+     *
+     * @param e          The thrown exception
+     * @param msg Method from which the exception is thrown
+     */
+    private void handleException(String msg, Exception e) {
+        log.error(msg, e);
+    }
+
+    private int getInternalHTTPInboundEndpointPort(MessageContext synCtx) {
+        int portOffset;
+        int internalHttpApiPort = 0;
+
+        if (null != serverPortOffset) {
+            portOffset = Integer.parseInt(serverPortOffset);
+            internalHttpApiPort = Integer.parseInt(synCtx.getEnvironment().getSynapseConfiguration().
```

Comment:
Does the INTERNAL_HTTP_API_PORT change after server startup? Do we need to get the info for every request?

Reply:
It depends on the Synapse Context.

Reply:
This won't change in runtime

Reply:
Changed the logic to read  INTERNAL_HTTP_API_PORT from the synapse properties.